File: android/pretext-core/src/main/kotlin/io/github/matrixy/pretext/Pretext.kt
package io.github.matrixy.pretext

import io.github.matrixy.pretext.analysis.*
import io.github.matrixy.pretext.bidi.computeSegmentLevels
import io.github.matrixy.pretext.linebreak.*
import io.github.matrixy.pretext.measurement.sharedMeasurementCache

/**
 * Entry point for the Pretext text layout engine.
 *
 * All public layout operations are accessed through this singleton. The typical workflow is:
 *
 * 1. Register a [TextSegmenter] once via [setSegmenter].
 * 2. Call [prepare] (fast path) or [prepareWithSegments] (rich path) to analyze and measure the text.
 * 3. Call [layout], [layoutWithLines], [walkLineRanges], or [layoutNextLine] at any width without
 *    further measurement cost.
 *
 * ```kotlin
 * Pretext.setSegmenter(IcuTextSegmenter())
 * val prepared = Pretext.prepare(text, PaintTextMeasurer(paint))
 * val result = Pretext.layout(prepared, maxWidth = 300f, lineHeight = 20f)
 * ```
 */
object Pretext {
    @Volatile
    private var segmenter: TextSegmenter? = null

    /**
     * Registers the [TextSegmenter] used by all subsequent [prepare] and [prepareWithSegments] calls.
     *
     * Must be called at least once before any preparation call. On Android, use
     * [io.github.matrixy.pretext.android.IcuTextSegmenter].
     *
     * @param segmenter The segmenter implementation to use.
     * @see TextSegmenter
     */
    fun setSegmenter(segmenter: TextSegmenter) {
        this.segmenter = segmenter
    }

    /**
     * Prepares text for fast layout by running the two-phase analysis+measurement pipeline.
     *
     * This is the **opaque fast-path** handle. It normalizes, segments, and measures the text
     * once; the returned [PreparedText] can then be laid out at any width via [layout] with no
     * further string work or canvas calls.
     *
     * Call this once per (text, font) pair and reuse the result across multiple widths.
     *
     * @param text The input string to prepare.
     * @param measurer A [TextMeasurer] configured for the target font and size.
     * @param options Optional [PrepareOptions] (e.g., whitespace mode).
     * @return An opaque [PreparedText] handle for use with [layout].
     * @throws IllegalStateException if [setSegmenter] has not been called.
     * @see layout
     */
    fun prepare(text: String, measurer: TextMeasurer, options: PrepareOptions = PrepareOptions()): PreparedText {
        val seg = requireNotNull(segmenter) { "Call Pretext.setSegmenter() before prepare()" }
        val analysis = analyzeText(text, EngineProfile.DEFAULT, options.whiteSpace, seg)
        return PreparedText(measureAnalysis(analysis, measurer, seg, includeSegments = false).core)
    }

    /**
     * Prepares text with the full segment array retained for rich layout APIs.
     *
     * Like [prepare], but the returned [PreparedTextWithSegments] also carries the segment
     * strings needed by [layoutWithLines], [walkLineRanges], and [layoutNextLine].
     *
     * Use this variant when you need line text content or cursor-based iteration, not just
     * a line count.
     *
     * @param text The input string to prepare.
     * @param measurer A [TextMeasurer] configured for the target font and size.
     * @param options Optional [PrepareOptions].
     * @return A [PreparedTextWithSegments] handle for use with the rich layout APIs.
     * @throws IllegalStateException if [setSegmenter] has not been called.
     * @see layoutWithLines
     * @see walkLineRanges
     * @see layoutNextLine
     */
    fun prepareWithSegments(text: String, measurer: TextMeasurer, options: PrepareOptions = PrepareOptions()): PreparedTextWithSegments {
        val seg = requireNotNull(segmenter) { "Call Pretext.setSegmenter() before prepare()" }
        val analysis = analyzeText(text, EngineProfile.DEFAULT, options.whiteSpace, seg)
        val result = measureAnalysis(analysis, measurer, seg, includeSegments = true)
        return PreparedTextWithSegments(result.core, result.segments ?: emptyArray())
    }

    /**
     * Computes line count and total height for the given width. **This is the resize hot path.**
     *
     * No string work, no canvas calls, no allocations beyond the result object. Safe to call
     * on every frame or constraint change.
     *
     * @param prepared The opaque handle from [prepare].
     * @param maxWidth Available width in pixels.
     * @param lineHeight Line height in pixels (uniform for all lines).
     * @return A [LayoutResult] with line count and total height.
     */
    fun layout(prepared: PreparedText, maxWidth: Float, lineHeight: Float): LayoutResult {
        val lineCount = countPreparedLines(prepared.core, maxWidth)
        return LayoutResult(lineCount, lineCount * lineHeight)
    }

    /**
     * Lays out text and materializes every line's text content, width, and cursor boundaries.
     *
     * Use this when you need the actual string content of each line (e.g., for custom rendering
     * or text selection). For aggregate geometry without string materialization, prefer
     * [walkLineRanges].
     *
     * @param prepared The rich handle from [prepareWithSegments].
     * @param maxWidth Available width in pixels.
     * @param lineHeight Line height in pixels.
     * @return A [LayoutLinesResult] containing all materialized lines.
     * @see walkLineRanges
     */
    fun layoutWithLines(prepared: PreparedTextWithSegments, maxWidth: Float, lineHeight: Float): LayoutLinesResult {
        if (prepared.core.widths.isEmpty()) return LayoutLinesResult(0, 0f, emptyList())
        val lines = mutableListOf<LayoutLine>()
        val lineCount = walkPreparedLines(prepared.core, maxWidth) { line ->
            lines.add(materializeLayoutLine(prepared, line))
        }
        return LayoutLinesResult(lineCount, lineCount * lineHeight, lines)
    }

    /**
     * Walks all lines without materializing their text, invoking [onLine] for each.
     *
     * This is the preferred batch geometry API when you need line widths and cursor
     * boundaries but not the line strings themselves (e.g., shrinkwrap calculations,
     * height pre-computation, or aggregate layout work).
     *
     * @param prepared The rich handle from [prepareWithSegments].
     * @param maxWidth Available width in pixels.
     * @param onLine Callback invoked for each line with its [LayoutLineRange].
     * @return The total number of lines.
     */
    fun walkLineRanges(prepared: PreparedTextWithSegments, maxWidth: Float, onLine: (LayoutLineRange) -> Unit): Int {
        if (prepared.core.widths.isEmpty()) return 0
        return walkPreparedLines(prepared.core, maxWidth) { line ->
            onLine(toLayoutLineRange(line))
        }
    }

    /**
     * Lays out a single line starting from [start] and returns it, or `null` if no content remains.
     *
     * This is the variable-width iterator for custom layout scenarios where each line may have
     * a different available width (e.g., text flowing around obstacles). Advance the cursor by
     * using the returned line's [LayoutLine.end] as the next call's [start].
     *
     * ```kotlin
     * var cursor = LayoutCursor(0, 0)
     * while (true) {
     *     val line = Pretext.layoutNextLine(prepared, cursor, currentWidth) ?: break
     *     // render line
     *     cursor = line.end
     * }
     * ```
     *
     * @param prepared The rich handle from [prepareWithSegments].
     * @param start Cursor position to begin the line from.
     * @param maxWidth Available width in pixels for this line.
     * @return The next [LayoutLine], or `null` if all content has been consumed.
     */
    fun layoutNextLine(prepared: PreparedTextWithSegments, start: LayoutCursor, maxWidth: Float): LayoutLine? {
        val range = layoutNextLineRangePublic(prepared.core, start, maxWidth) ?: return null
        return materializeLine(prepared, toLayoutLineRange(range))
    }

    /**
     * Clears the shared segment-metrics measurement cache.
     *
     * Call this when the font configuration changes or to reclaim memory. The cache is
     * structured as `Map<font, Map<segment, metrics>>` and is shared across all prepared texts.
     */
    fun clearCache() {
        sharedMeasurementCache.clear()
    }

    /**
     * Sets the locale used for future word segmentation and clears the measurement cache.
     *
     * Pass `null` to revert to the runtime default locale. This affects how
     * [TextSegmenter] instances created by the platform binding resolve word boundaries.
     *
     * @param locale The desired locale, or `null` for the system default.
     */
    fun setLocale(locale: java.util.Locale?) {
        localeOverride = locale
        clearCache()
    }

    // Stored locale for segmenter creation
    @Volatile
    internal var localeOverride: java.util.Locale? = null
}

// Internal orchestration
internal data class MeasureResult(val core: PreparedCore, val segments: Array<String>?)

internal fun measureAnalysis(
    analysis: TextAnalysis,
    measurer: TextMeasurer,
    segmenter: TextSegmenter,
    includeSegments: Boolean
): MeasureResult {
    val cache = sharedMeasurementCache
    val profile = EngineProfile.DEFAULT
    val fontKey = measurer // Use the measurer itself as font key

    val hyphenWidth = cache.getMetrics("-", fontKey, measurer).width
    val spaceWidth = cache.getMetrics(" ", fontKey, measurer).width
    val tabStopAdvance = spaceWidth * 8

    if (analysis.len == 0) {
        return MeasureResult(
            PreparedCore(
                FloatArray(0), FloatArray(0), FloatArray(0),
                emptyArray(), true, null,
                emptyArray(), emptyArray(),
                hyphenWidth, tabStopAdvance, emptyArray()
            ),
            if (includeSegments) emptyArray() else null
        )
    }

    val widths = mutableListOf<Float>()
    val lineEndFitAdvances = mutableListOf<Float>()
    val lineEndPaintAdvances = mutableListOf<Float>()
    val kinds = mutableListOf<SegmentBreakKind>()
    var simpleLineWalkFastPath = analysis.chunks.size <= 1
    val segStarts = if (includeSegments) mutableListOf<Int>() else null
    val breakableWidths = mutableListOf<FloatArray?>()
    val breakablePrefixWidths = mutableListOf<FloatArray?>()
    val segments = if (includeSegments) mutableListOf<String>() else null
    val preparedStartByAnalysisIndex = IntArray(analysis.len)
    val preparedEndByAnalysisIndex = IntArray(analysis.len)

    fun pushMeasuredSegment(
        text: String, width: Float, fitAdv: Float, paintAdv: Float,
        kind: SegmentBreakKind, start: Int,
        breakable: FloatArray?, breakablePrefix: FloatArray?
    ) {
        if (kind != SegmentBreakKind.Text && kind != SegmentBreakKind.Space && kind != SegmentBreakKind.ZeroWidthBreak) {
            simpleLineWalkFastPath = false
        }
        widths.add(width)
        lineEndFitAdvances.add(fitAdv)
        lineEndPaintAdvances.add(paintAdv)
        kinds.add(kind)
        segStarts?.add(start)
        breakableWidths.add(breakable)
        breakablePrefixWidths.add(breakablePrefix)
        segments?.add(text)
    }

    for (mi in 0 until analysis.len) {
        preparedStartByAnalysisIndex[mi] = widths.size
        val segText = analysis.texts[mi]
        val segWordLike = analysis.isWordLike[mi]
        val segKind = analysis.kinds[mi]
        val segStart = analysis.starts[mi]

        when (segKind) {
            SegmentBreakKind.SoftHyphen -> {
                pushMeasuredSegment(segText, 0f, hyphenWidth, hyphenWidth, segKind, segStart, null, null)
                preparedEndByAnalysisIndex[mi] = widths.size; continue
            }
            SegmentBreakKind.HardBreak, SegmentBreakKind.Tab -> {
                pushMeasuredSegment(segText, 0f, 0f, 0f, segKind, segStart, null, null)
                preparedEndByAnalysisIndex[mi] = widths.size; continue
            }
            else -> {}
        }

        val segMetrics = cache.getMetrics(segText, fontKey, measurer)

        // CJK grapheme splitting
        if (segKind == SegmentBreakKind.Text && segMetrics.containsCJK) {
            val graphemes = segmenter.segmentGraphemes(segText)
            var unitText = ""
            for (grapheme in graphemes) {
                if (unitText.isEmpty()) { unitText = grapheme; continue }

                val unitFirst = unitText[0]
                val graphFirst = grapheme[0]
                if (kinsokuEnd.contains(unitFirst) || kinsokuStart.contains(graphFirst) ||
                    leftStickyPunctuation.contains(graphFirst) ||
                    (profile.carryCJKAfterClosingQuote && isCJK(grapheme) && endsWithClosingQuote(unitText))) {
                    unitText += grapheme; continue
                }

                val unitMetrics = cache.getMetrics(unitText, fontKey, measurer)
                pushMeasuredSegment(unitText, unitMetrics.width, unitMetrics.width, unitMetrics.width, SegmentBreakKind.Text, segStart, null, null)
                unitText = grapheme
            }
            if (unitText.isNotEmpty()) {
                val unitMetrics = cache.getMetrics(unitText, fontKey, measurer)
                pushMeasuredSegment(unitText, unitMetrics.width, unitMetrics.width, unitMetrics.width, SegmentBreakKind.Text, segStart, null, null)
            }
            preparedEndByAnalysisIndex[mi] = widths.size; continue
        }

        val w = segMetrics.width
        val lineEndFitAdv = if (segKind == SegmentBreakKind.Space || segKind == SegmentBreakKind.PreservedSpace || segKind == SegmentBreakKind.ZeroWidthBreak) 0f else w
        val lineEndPaintAdv = if (segKind == SegmentBreakKind.Space || segKind == SegmentBreakKind.ZeroWidthBreak) 0f else w

        if (segWordLike && segText.length > 1) {
            val gWidths = cache.getGraphemeWidths(segText, fontKey, measurer, segmenter)
            val gPrefixWidths = if (profile.preferPrefixWidthsForBreakableRuns) cache.getGraphemePrefixWidths(segText, fontKey, measurer, segmenter) else null
            pushMeasuredSegment(segText, w, lineEndFitAdv, lineEndPaintAdv, segKind, segStart, gWidths, gPrefixWidths)
        } else {
            pushMeasuredSegment(segText, w, lineEndFitAdv, lineEndPaintAdv, segKind, segStart, null, null)
        }
        preparedEndByAnalysisIndex[mi] = widths.size
    }

    val chunks = mapAnalysisChunksToPreparedChunks(analysis.chunks, preparedStartByAnalysisIndex, preparedEndByAnalysisIndex)
    val segLevels = segStarts?.let { computeSegmentLevels(analysis.normalized, it.toIntArray()) }

    return MeasureResult(
        PreparedCore(
            widths.toFloatArray(), lineEndFitAdvances.toFloatArray(), lineEndPaintAdvances.toFloatArray(),
            kinds.toTypedArray(), simpleLineWalkFastPath, segLevels,
            breakableWidths.toTypedArray(), breakablePrefixWidths.toTypedArray(),
            hyphenWidth, tabStopAdvance, chunks
        ),
        segments?.toTypedArray()
    )
}

private fun mapAnalysisChunksToPreparedChunks(
    chunks: List<AnalysisChunk>,
    startMap: IntArray,
    endMap: IntArray
): Array<PreparedLineChunk> {
    return Array(chunks.size) { i ->
        val chunk = chunks[i]
        val start = if (chunk.startSegmentIndex < startMap.size) startMap[chunk.startSegmentIndex] else (endMap.lastOrNull() ?: 0)
        val end = if (chunk.endSegmentIndex < startMap.size) startMap[chunk.endSegmentIndex] else (endMap.lastOrNull() ?: 0)
        val consumed = if (chunk.consumedEndSegmentIndex < startMap.size) startMap[chunk.consumedEndSegmentIndex] else (endMap.lastOrNull() ?: 0)
        PreparedLineChunk(start, end, consumed)
    }
}

// Line materialization helpers
internal fun toLayoutLineRange(line: InternalLayoutLine): LayoutLineRange {
    return LayoutLineRange(
        line.width,
        LayoutCursor(line.startSegmentIndex, line.startGraphemeIndex),
        LayoutCursor(line.endSegmentIndex, line.endGraphemeIndex)
    )
}

private fun lineHasDiscretionaryHyphen(kinds: Array<SegmentBreakKind>, startSeg: Int, startGrapheme: Int, endSeg: Int): Boolean {
    return endSeg > 0 && kinds[endSeg - 1] == SegmentBreakKind.SoftHyphen && !(startSeg == endSeg && startGrapheme > 0)
}

// Grapheme cluster segmentation using java.text.BreakIterator (JVM-safe, no Android dep)
private fun splitGraphemeClusters(text: String): List<String> {
    val bi = java.text.BreakIterator.getCharacterInstance()
    bi.setText(text)
    val graphemes = mutableListOf<String>()
    var start = bi.first()
    var end = bi.next()
    while (end != java.text.BreakIterator.DONE) {
        graphemes.add(text.substring(start, end))
        start = end
        end = bi.next()
    }
    return graphemes
}

private fun buildLineText(
    segments: Array<String>, kinds: Array<SegmentBreakKind>,
    startSeg: Int, startGrapheme: Int, endSeg: Int, endGrapheme: Int
): String {
    val sb = StringBuilder()
    val hasHyphen = lineHasDiscretionaryHyphen(kinds, startSeg, startGrapheme, endSeg)

    for (i in startSeg until endSeg) {
        if (kinds[i] == SegmentBreakKind.SoftHyphen || kinds[i] == SegmentBreakKind.HardBreak) continue
        if (i == startSeg && startGrapheme > 0) {
            val graphemes = splitGraphemeClusters(segments[i])
            sb.append(graphemes.drop(startGrapheme).joinToString(""))
        } else {
            sb.append(segments[i])
        }
    }

    if (endGrapheme > 0) {
        if (hasHyphen) sb.append('-')
        val graphemes = splitGraphemeClusters(segments[endSeg])
        val from = if (startSeg == endSeg) startGrapheme else 0
        sb.append(graphemes.subList(from, endGrapheme).joinToString(""))
    } else if (hasHyphen) {
        sb.append('-')
    }

    return sb.toString()
}

internal fun materializeLayoutLine(prepared: PreparedTextWithSegments, line: InternalLayoutLine): LayoutLine {
    return LayoutLine(
        buildLineText(prepared.segments, prepared.core.kinds,
            line.startSegmentIndex, line.startGraphemeIndex,
            line.endSegmentIndex, line.endGraphemeIndex),
        line.width,
        LayoutCursor(line.startSegmentIndex, line.startGraphemeIndex),
        LayoutCursor(line.endSegmentIndex, line.endGraphemeIndex)
    )
}

internal fun materializeLine(prepared: PreparedTextWithSegments, range: LayoutLineRange): LayoutLine {
    return LayoutLine(
        buildLineText(prepared.segments, prepared.core.kinds,
            range.start.segmentIndex, range.start.graphemeIndex,
            range.end.segmentIndex, range.end.graphemeIndex),
        range.width,
        range.start,
        range.end
    )
}

File: ios/Sources/Pretext/Pretext.swift
import Foundation

// MARK: - Public API

/// Pure-arithmetic text measurement and layout engine.
///
/// `Pretext` is a namespace (caseless enum) that exposes a two-phase API:
///
/// 1. **Prepare** — analyze and measure text once using CoreText.
/// 2. **Layout** — compute line breaks and height using only cached arithmetic.
///
/// The prepare phase is O(segments) with CoreText calls; the layout phase is
/// sub-microsecond per call with no font or canvas work.
///
/// ```swift
/// let font = FontSpec(name: "Helvetica Neue", size: 16)
/// let prepared = Pretext.prepare("Hello, world!", font: font)
/// let result = Pretext.layout(prepared, maxWidth: 300, lineHeight: 22)
/// ```
public enum Pretext {

    /// Analyzes and measures text, returning an opaque handle for fast layout.
    ///
    /// This is the recommended entry point for most use cases. The returned
    /// ``PreparedText`` caches segment widths and break metadata so that
    /// subsequent ``layout(_:maxWidth:lineHeight:)`` calls are pure arithmetic.
    ///
    /// Call once per text/font combination; call ``layout(_:maxWidth:lineHeight:)``
    /// on every width change.
    ///
    /// - Parameters:
    ///   - text: The input string. Supports all scripts, emoji, and bidirectional text.
    ///   - font: Font descriptor for CoreText measurement.
    ///   - options: Whitespace mode and other analysis options.
    /// - Returns: An opaque prepared handle.
    public static func prepare(_ text: String, font: FontSpec, options: PrepareOptions = PrepareOptions()) -> PreparedText {
        let analysis = analyzeText(text, profile: .default, whiteSpace: options.whiteSpace)
        return PreparedText(core: measureAnalysis(analysis, font: font, includeSegments: false).core)
    }

    /// Analyzes and measures text, returning a rich handle that includes segment strings.
    ///
    /// Use this variant when you need line-level text content via
    /// ``layoutWithLines(_:maxWidth:lineHeight:)``, ``layoutNextLine(_:start:maxWidth:)``,
    /// or ``walkLineRanges(_:maxWidth:onLine:)``.
    ///
    /// - Parameters:
    ///   - text: The input string.
    ///   - font: Font descriptor for CoreText measurement.
    ///   - options: Whitespace mode and other analysis options.
    /// - Returns: A prepared handle with accessible segment data.
    public static func prepareWithSegments(_ text: String, font: FontSpec, options: PrepareOptions = PrepareOptions()) -> PreparedTextWithSegments {
        let analysis = analyzeText(text, profile: .default, whiteSpace: options.whiteSpace)
        let result = measureAnalysis(analysis, font: font, includeSegments: true)
        return PreparedTextWithSegments(core: result.core, segments: result.segments ?? [])
    }

    /// Computes line count and total height — the hot path.
    ///
    /// This is pure arithmetic over cached segment widths: no CoreText calls, no string
    /// work, no allocations. Safe to call on every frame or resize event.
    ///
    /// - Parameters:
    ///   - prepared: A handle from ``prepare(_:font:options:)``.
    ///   - maxWidth: Available horizontal space in points.
    ///   - lineHeight: Fixed line height in points.
    /// - Returns: Line count and total height.
    public static func layout(_ prepared: PreparedText, maxWidth: Double, lineHeight: Double) -> LayoutResult {
        let lineCount = countPreparedLines(prepared.core, maxWidth: maxWidth)
        return LayoutResult(lineCount: lineCount, height: Double(lineCount) * lineHeight)
    }

    /// Computes layout and returns every line's text, width, and boundary cursors.
    ///
    /// Use when you need the actual string content of each line (e.g. for custom rendering).
    /// Requires a ``PreparedTextWithSegments`` from ``prepareWithSegments(_:font:options:)``.
    ///
    /// - Parameters:
    ///   - prepared: A rich prepared handle.
    ///   - maxWidth: Available horizontal space in points.
    ///   - lineHeight: Fixed line height in points.
    /// - Returns: Line count, total height, and an array of materialized lines.
    public static func layoutWithLines(_ prepared: PreparedTextWithSegments, maxWidth: Double, lineHeight: Double) -> LayoutLinesResult {
        var lines: [LayoutLine] = []
        guard !prepared.core.widths.isEmpty else {
            return LayoutLinesResult(lineCount: 0, height: 0, lines: [])
        }

        let lineCount = walkPreparedLines(prepared.core, maxWidth: maxWidth) { internalLine in
            lines.append(materializeLayoutLine(prepared, line: internalLine))
        }

        return LayoutLinesResult(lineCount: lineCount, height: Double(lineCount) * lineHeight, lines: lines)
    }

    /// Walks all lines without materializing text strings.
    ///
    /// Calls `onLine` for each line with its width and boundary cursors. This is the
    /// non-materializing batch API — useful for aggregate geometry work like shrinkwrap
    /// or hit-testing where you need line widths but not the text content.
    ///
    /// - Parameters:
    ///   - prepared: A rich prepared handle.
    ///   - maxWidth: Available horizontal space in points.
    ///   - onLine: Callback invoked once per line with a ``LayoutLineRange``.
    /// - Returns: Total number of lines.
    @discardableResult
    public static func walkLineRanges(_ prepared: PreparedTextWithSegments, maxWidth: Double, onLine: @escaping (LayoutLineRange) -> Void) -> Int {
        guard !prepared.core.widths.isEmpty else { return 0 }
        return walkPreparedLines(prepared.core, maxWidth: maxWidth) { line in
            onLine(toLayoutLineRange(line))
        }
    }

    /// Lays out a single line starting from the given cursor, for variable-width iteration.
    ///
    /// Use this to flow text around obstacles or into columns with varying widths.
    /// Pass a different `maxWidth` on each call. Returns `nil` when the text is exhausted.
    ///
    /// ```swift
    /// var cursor = LayoutCursor(segmentIndex: 0, graphemeIndex: 0)
    /// while let line = Pretext.layoutNextLine(prepared, start: cursor, maxWidth: width) {
    ///     cursor = line.end
    /// }
    /// ```
    ///
    /// - Parameters:
    ///   - prepared: A rich prepared handle.
    ///   - start: Cursor position to begin the line from.
    ///   - maxWidth: Available horizontal space for this line.
    /// - Returns: The next materialized line, or `nil` if all text has been consumed.
    public static func layoutNextLine(_ prepared: PreparedTextWithSegments, start: LayoutCursor, maxWidth: Double) -> LayoutLine? {
        let range = layoutNextLineRangePublic(prepared.core, start: start, maxWidth: maxWidth)
        guard let range = range else { return nil }
        return materializeLine(prepared, range: range)
    }

    /// Clears the shared segment metrics cache and resets the word segmenter locale.
    ///
    /// Call after a font change or when memory pressure requires releasing cached measurements.
    /// Subsequent `prepare` calls will re-measure from scratch.
    public static func clearCache() {
        sharedMeasurementCache.clear()
        setSegmenterLocale(nil)
    }

    /// Sets the locale used by the word segmenter for future `prepare` calls.
    ///
    /// Different locales can produce different word boundaries (e.g. Thai, Lao, Khmer).
    /// Pass `nil` to revert to the system default. Also clears the measurement cache,
    /// since segment boundaries may change.
    ///
    /// - Parameter locale: The locale to use, or `nil` for the system default.
    public static func setLocale(_ locale: Locale?) {
        setSegmenterLocale(locale)
        sharedMeasurementCache.clear()
    }
}

// MARK: - Internal Orchestration

struct MeasureResult {
    let core: PreparedCore
    let segments: [String]?
}

func measureAnalysis(_ analysis: TextAnalysis, font: FontSpec, includeSegments: Bool) -> MeasureResult {
    let measurer = CoreTextMeasurer(font: font)
    let cache = sharedMeasurementCache
    let profile = EngineProfile.default

    let hyphenWidth = cache.getMetrics(segment: "-", font: font, measurer: measurer).width
    let spaceWidth = cache.getMetrics(segment: " ", font: font, measurer: measurer).width
    let tabStopAdvance = spaceWidth * 8

    guard analysis.len > 0 else {
        return MeasureResult(
            core: PreparedCore(
                widths: [], lineEndFitAdvances: [], lineEndPaintAdvances: [],
                kinds: [], simpleLineWalkFastPath: true, segLevels: nil,
                breakableWidths: [], breakablePrefixWidths: [],
                discretionaryHyphenWidth: hyphenWidth, tabStopAdvance: tabStopAdvance,
                chunks: []
            ),
            segments: includeSegments ? [] : nil
        )
    }

    var widths: [Double] = []
    var lineEndFitAdvances: [Double] = []
    var lineEndPaintAdvances: [Double] = []
    var kinds: [SegmentBreakKind] = []
    var simpleLineWalkFastPath = analysis.chunks.count <= 1
    var segStarts: [Int]? = includeSegments ? [] : nil
    var breakableWidths: [[Double]?] = []
    var breakablePrefixWidths: [[Double]?] = []
    var segments: [String]? = includeSegments ? [] : nil
    var preparedStartByAnalysisIndex = [Int](repeating: 0, count: analysis.len)
    var preparedEndByAnalysisIndex = [Int](repeating: 0, count: analysis.len)

    func pushMeasuredSegment(_ text: String, _ width: Double, _ fitAdv: Double, _ paintAdv: Double,
                             _ kind: SegmentBreakKind, _ start: Int,
                             _ breakable: [Double]?, _ breakablePrefix: [Double]?) {
        if kind != .text && kind != .space && kind != .zeroWidthBreak {
            simpleLineWalkFastPath = false
        }
        widths.append(width)
        lineEndFitAdvances.append(fitAdv)
        lineEndPaintAdvances.append(paintAdv)
        kinds.append(kind)
        segStarts?.append(start)
        breakableWidths.append(breakable)
        breakablePrefixWidths.append(breakablePrefix)
        segments?.append(text)
    }

    for mi in 0..<analysis.len {
        preparedStartByAnalysisIndex[mi] = widths.count
        let segText = analysis.texts[mi]
        let segWordLike = analysis.isWordLike[mi]
        let segKind = analysis.kinds[mi]
        let segStart = analysis.starts[mi]

        if segKind == .softHyphen {
            pushMeasuredSegment(segText, 0, hyphenWidth, hyphenWidth, segKind, segStart, nil, nil)
            preparedEndByAnalysisIndex[mi] = widths.count
            continue
        }

        if segKind == .hardBreak {
            pushMeasuredSegment(segText, 0, 0, 0, segKind, segStart, nil, nil)
            preparedEndByAnalysisIndex[mi] = widths.count
            continue
        }

        if segKind == .tab {
            pushMeasuredSegment(segText, 0, 0, 0, segKind, segStart, nil, nil)
            preparedEndByAnalysisIndex[mi] = widths.count
            continue
        }

        let segMetrics = cache.getMetrics(segment: segText, font: font, measurer: measurer)

        // CJK grapheme splitting
        if segKind == .text && segMetrics.containsCJK {
            var unitText = ""
            var unitStart = segStart

            let graphemes = segmentGraphemes(segText)
            var graphemeOffset = 0
            for grapheme in graphemes {
                if unitText.isEmpty {
                    unitText = grapheme
                    unitStart = segStart + graphemeOffset
                    graphemeOffset += grapheme.utf16.count
                    continue
                }

                let graphemeChar = grapheme.first!
                let unitFirstChar = unitText.first!

                if kinsokuEnd.contains(unitFirstChar) ||
                   kinsokuStart.contains(graphemeChar) ||
                   leftStickyPunctuation.contains(graphemeChar) ||
                   (profile.carryCJKAfterClosingQuote && isCJK(grapheme) && endsWithClosingQuote(unitText)) {
                    unitText += grapheme
                    graphemeOffset += grapheme.utf16.count
                    continue
                }

                let unitMetrics = cache.getMetrics(segment: unitText, font: font, measurer: measurer)
                pushMeasuredSegment(unitText, unitMetrics.width, unitMetrics.width, unitMetrics.width, .text, unitStart, nil, nil)
                unitText = grapheme
                unitStart = segStart + graphemeOffset
                graphemeOffset += grapheme.utf16.count
            }

            if !unitText.isEmpty {
                let unitMetrics = cache.getMetrics(segment: unitText, font: font, measurer: measurer)
                pushMeasuredSegment(unitText, unitMetrics.width, unitMetrics.width, unitMetrics.width, .text, unitStart, nil, nil)
            }
            preparedEndByAnalysisIndex[mi] = widths.count
            continue
        }

        let w = segMetrics.width
        let lineEndFitAdvance: Double = (segKind == .space || segKind == .preservedSpace || segKind == .zeroWidthBreak) ? 0 : w
        let lineEndPaintAdvance: Double = (segKind == .space || segKind == .zeroWidthBreak) ? 0 : w

        if segWordLike && segText.count > 1 {
            let gWidths = cache.getGraphemeWidths(segment: segText, font: font, measurer: measurer)
            let gPrefixWidths = profile.preferPrefixWidthsForBreakableRuns
                ? cache.getGraphemePrefixWidths(segment: segText, font: font, measurer: measurer)
                : nil
            pushMeasuredSegment(segText, w, lineEndFitAdvance, lineEndPaintAdvance, segKind, segStart, gWidths, gPrefixWidths)
        } else {
            pushMeasuredSegment(segText, w, lineEndFitAdvance, lineEndPaintAdvance, segKind, segStart, nil, nil)
        }
        preparedEndByAnalysisIndex[mi] = widths.count
    }

    let preparedChunks = mapAnalysisChunksToPreparedChunks(analysis.chunks, preparedStartByAnalysisIndex, preparedEndByAnalysisIndex)
    let segLevels: [Int8]? = segStarts != nil ? computeSegmentLevels(normalized: analysis.normalized, segStarts: segStarts!) : nil

    return MeasureResult(
        core: PreparedCore(
            widths: widths, lineEndFitAdvances: lineEndFitAdvances, lineEndPaintAdvances: lineEndPaintAdvances,
            kinds: kinds, simpleLineWalkFastPath: simpleLineWalkFastPath, segLevels: segLevels,
            breakableWidths: breakableWidths, breakablePrefixWidths: breakablePrefixWidths,
            discretionaryHyphenWidth: hyphenWidth, tabStopAdvance: tabStopAdvance, chunks: preparedChunks
        ),
        segments: segments
    )
}

private func mapAnalysisChunksToPreparedChunks(
    _ chunks: [AnalysisChunk],
    _ startMap: [Int],
    _ endMap: [Int]
) -> [PreparedLineChunk] {
    chunks.map { chunk in
        let start = chunk.startSegmentIndex < startMap.count ? startMap[chunk.startSegmentIndex] : (endMap.last ?? 0)
        let end = chunk.endSegmentIndex < startMap.count ? startMap[chunk.endSegmentIndex] : (endMap.last ?? 0)
        let consumed = chunk.consumedEndSegmentIndex < startMap.count ? startMap[chunk.consumedEndSegmentIndex] : (endMap.last ?? 0)
        return PreparedLineChunk(startSegmentIndex: start, endSegmentIndex: end, consumedEndSegmentIndex: consumed)
    }
}

// MARK: - Line Materialization

private func toLayoutLineRange(_ line: InternalLayoutLine) -> LayoutLineRange {
    LayoutLineRange(
        width: line.width,
        start: LayoutCursor(segmentIndex: line.startSegmentIndex, graphemeIndex: line.startGraphemeIndex),
        end: LayoutCursor(segmentIndex: line.endSegmentIndex, graphemeIndex: line.endGraphemeIndex)
    )
}

private func lineHasDiscretionaryHyphen(_ kinds: [SegmentBreakKind], _ startSeg: Int, _ startGrapheme: Int, _ endSeg: Int) -> Bool {
    endSeg > 0 && kinds[endSeg - 1] == .softHyphen && !(startSeg == endSeg && startGrapheme > 0)
}

private func buildLineText(_ segments: [String], _ kinds: [SegmentBreakKind],
                            _ startSeg: Int, _ startGrapheme: Int,
                            _ endSeg: Int, _ endGrapheme: Int) -> String {
    var text = ""
    let hasHyphen = lineHasDiscretionaryHyphen(kinds, startSeg, startGrapheme, endSeg)

    for i in startSeg..<endSeg {
        if kinds[i] == .softHyphen || kinds[i] == .hardBreak { continue }
        if i == startSeg && startGrapheme > 0 {
            let graphemes = segmentGraphemes(segments[i])
            text += graphemes[startGrapheme...].joined()
        } else {
            text += segments[i]
        }
    }

    if endGrapheme > 0 {
        if hasHyphen { text += "-" }
        let graphemes = segmentGraphemes(segments[endSeg])
        let from = (startSeg == endSeg) ? startGrapheme : 0
        text += graphemes[from..<endGrapheme].joined()
    } else if hasHyphen {
        text += "-"
    }

    return text
}

private func materializeLayoutLine(_ prepared: PreparedTextWithSegments, line: InternalLayoutLine) -> LayoutLine {
    LayoutLine(
        text: buildLineText(prepared.segments, prepared.core.kinds,
                            line.startSegmentIndex, line.startGraphemeIndex,
                            line.endSegmentIndex, line.endGraphemeIndex),
        width: line.width,
        start: LayoutCursor(segmentIndex: line.startSegmentIndex, graphemeIndex: line.startGraphemeIndex),
        end: LayoutCursor(segmentIndex: line.endSegmentIndex, graphemeIndex: line.endGraphemeIndex)
    )
}

private func materializeLine(_ prepared: PreparedTextWithSegments, range: LayoutLineRange) -> LayoutLine {
    LayoutLine(
        text: buildLineText(prepared.segments, prepared.core.kinds,
                            range.start.segmentIndex, range.start.graphemeIndex,
                            range.end.segmentIndex, range.end.graphemeIndex),
        width: range.width,
        start: range.start,
        end: range.end
    )
}
